File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode temp = head;
        ListNode prev = null;
        ListNode front = null;
        int count = 0;

        // First, check if there are at least k nodes left to reverse
        ListNode check = head;
        for (int i = 0; i < k; i++) {
            if (check == null){
                return head;
            }
            check = check.next;
        }

        // Reverse k nodes
        while (temp != null && count < k) {
            front = temp.next;
            temp.next = prev;
            prev = temp;
            temp = front;
            count++;
        }

        // Recursively call reverseKGroup for the next segment
        if (front != null) {
            head.next = reverseKGroup(front, k);
        }

        return prev;
    }
}

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head or k == 1:
            return head

        curr = head
        new_head = None
        tail = None

        while curr:
            # Step 1: Check if there are at least k nodes
            temp = curr
            count = 0
            while temp and count < k:
                temp = temp.next
                count += 1

            if count < k:
                if tail:
                    tail.next = curr
                break

            # Step 2: Reverse k nodes
            group_head = curr
            prev = None
            next_node = None
            count = 0

            while curr and count < k:
                next_node = curr.next
                curr.next = prev
                prev = curr
                curr = next_node
                count += 1

            # Step 3: Connect reversed group
            if not new_head:
                new_head = prev
            if tail:
                tail.next = prev
            tail = group_head

        return new_head if new_head else head